flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -594,6 +594,25 @@ class ScalarExprLowering {
     fir::emitFatalError(getLoc(), "NULL() must be lowered in its context");
   }
 
+  /// A `NULL()` in a position where a mutable box is expected has the same
+  /// semantics as an absent optional box value.
```

Comment:
It bothers me a little bit to deal with the call context lowering of NULL() in `genMutableBoxValue` that is not specific to calls. If any of the other contexts dealing with pointer/allocatable miss the NULL special handling (described by Table 16.5 in the standard) and fall here, we might be up for weird runtime bugs (especially with the `(!fir.ref<!fir.llvm_ptr<none>>) -> !fir.ref<!fir.box<none>>` trick).
Now, I think the other places dealing with `NULL()` are covered and this change seems functional, so I am ok to try making this safer later.

Reply:
I tried several ideas here, but there are a lot of assertions and implicit assumptions about how the IR will be presented going on. A null pointer (LLVM sense) with no other information can't really be turned into a real box (no type, no rank, etc., etc.). But everything up to this is expecting (and asserting) that there will be a reference to a box. Without changing those underlying assumptions, the most straightforward approach appears to be creating a degenerate box around the null pointer. This solution does create a viable box variable (mutable box value), so that seems ok. The variable just has the degenerate null pointer.

Agree that this area may need some more revision/clean up.